File: DoSemestralka/Batoh.cpp
#include <iostream>
#include <string>

#include "Batoh.h"
#include "BatohItem.h"


using namespace std;

Batoh::Batoh()
{
}


Batoh::~Batoh()
{
}

void Batoh::AddItem(BatohItem * item)
{
	if (item != nullptr)
	{
		_actualCapacity = item->GetVaha();
		_hodnotaBatohu = item->GetCena();
		_items.push_back(item);

		std::cout << "Loaded " << item->ToString() << endl;
	} else
	{
		cout << "Tried to insert invalid nullptr" << endl;
	}
	
}

void Batoh::RemoveTheHardestItem()
{
	if (_items.empty() == false)
	{
		BatohItem * theHardest = _items.front();
		int theHardestIndex = 0;
		
		int tmpIndex = 0;

		for (auto const tmp : _items)
		{
			if (tmp->GetCena() > theHardest->GetVaha())
			{
				theHardest = tmp;
				theHardestIndex = tmpIndex;
			}
			tmpIndex++;
		}
		if (theHardest != nullptr)
		{
			std::cout << "Removing from bag" << theHardest->ToString() << std::endl;
			_items.erase(_items.begin() + tmpIndex);
		}

	}

}

int Batoh::GetMaxCapacity()
{
	return _maxCapacity;
}

int Batoh::GetActualCapacity()
{
	return _actualCapacity;
}

int Batoh::GetFreeCapacity() 
{
	return _maxCapacity - _actualCapacity;
}

int Batoh::GetHodnotaBatohu() 
{
	return _hodnotaBatohu;
}

void Batoh::SetMaxCapacity(int capacity)
{
	_maxCapacity = capacity;
}

void Batoh::PrintInventory()
{
	std::cout << "Total bag weight: ";
	std::cout << _actualCapacity;

	std::cout << "Free space: ";
	std::cout << this->GetFreeCapacity();

	for (auto const tmp : _items)
	{
		std::cout << "Batoh contain: " << tmp->GetCena() << " with weight: " << tmp->GetVaha() << std::endl;
	}
}

File: DualnaVkladaciaHeuristika/src/dualnavkladaciaheuristika/Batoh.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package dualnavkladaciaheuristika;

import java.util.ArrayList;

/**
 *
 * @author Erik
 */
public class Batoh {
    
    private int _maxCapacity = 0;
    private int _actualCapacity = 0;
    private int _hodnotaBatohu = 0;
    
    private ArrayList<BatohItem> _items = new ArrayList();
    
    public void AddItem(BatohItem item)
    {
        if (item != null)
        {
            _items.add(item);
            _actualCapacity += item.GetVaha();
            _hodnotaBatohu += item.GetCena();
            
            //System.out.println("Added to bag " + item.toString());
            
        } else {
            System.out.println("Tried to insert null");
        }
        
    }
    
    public void RemoveTheHardestItem()
    {
        if (_items.isEmpty() == false)
        {
            BatohItem hardest = _items.get(0);
            int theHardestIndex = 0;
            
            int tmpIndex = 0;
            
            for (BatohItem tmpItem : _items)
            {
                if (tmpItem.GetVaha() > hardest.GetVaha())
                {
                    hardest = tmpItem;
                    theHardestIndex = tmpIndex;
                }
                tmpIndex++;
            }
            
            if (hardest != null)
            {
                _actualCapacity -= hardest.GetVaha();
                _hodnotaBatohu -= hardest.GetCena();
                
                System.out.println("Removing from bag " + hardest.toString());
                System.out.println("Actual bag capacity " + _actualCapacity);
                System.out.println("Hodnota batoha " + _hodnotaBatohu);
                System.out.println("---");
                
                _items.remove(theHardestIndex);
            }
            
            
        } 
    }
    
    public void SetMaximalCapacity(int capacity)
    {
        _maxCapacity = capacity;        
    }   
    
    public int GetMaxCapacity()
    {
        return _maxCapacity;
    }
    
    public int GetActualCapacity()
    {
        return _actualCapacity;
    }
    
    public int GetHodnotaBatohu()
    {
        return _hodnotaBatohu;
    }
    
    public int GetFreeCapacity()
    {
        return _maxCapacity - _actualCapacity;
    }
        
    public int GetPocetItemov()
    {
        return _items.size();
    }
    
    
    
    
    
    
    
}
